content/smyja/how-to-use-graphql-with-django.md

```diff
@@ -0,0 +1,287 @@
+
+ **Properties of Graphql**:
+
+The following terms would be used a lot when interacting with graphql. It is important that you know them, though we wouldn't cover all of them in this tutorial.
+
+- schema
+- query
+- Nesting
+- Mutation
+- Subscription
+- Resolvers
+
+**Schema:** 
+GraphQl Schema describes the functionality available to the client applications that connect to it. 
+
+**Query:**
+A Query is a type on a schema that defines the kind of operations that can be done to read or fetch data. Queries represent the `GET` request.
+
+**Resolvers:**
+They are functions that return values for fields that exist on types in a schema. 
+
+**Mutation:**
+A mutation is a type on a schema that defines the kind of operations that can be done to modify data.
+
+**Subscription:**
+Subscriptions in graphql are for notifying your client server in real time about changes in the data.
+
+**Nesting:**
+A nested query is a query that is a child of another query.
+
+## Setting up graphql with django
+First, we need to install Django and create a project. I will be using a bash script i created to do this.
+[Django-bash-script](https://github.com/Smyja/Django-bash)
+
+To use graphql with django, you will need to install graphene-django
+
+```pip install graphene-django```
+
+Add it to your settings.py file:
+```python
+INSTALLED_APPS = [
+    'graphene_django',
+
+    # ...
+]
+```
+If you run the server, you see this error:
+```ImportError: cannot import name 'force_text' from 'django.utils.encoding```
+
+add this to your settings.py file:
+
+```python
+import django
+from django.utils.encoding import force_str
+
+django.utils.encoding.force_text = force_str
+
+```
+or downgrade your django version.
+
+Create a Models.py file in your project that contains tthe models you want to use.:
+```python
+from django.db import models
+class Restaurant(models.Model):
+    name = models.CharField(max_length=100)
+    address = models.CharField(max_length=200)
+
+    def __str__(self):
+        return self.name
+```
+
+
+Add a graphql route to your urls.py file for django 2.0 and above:
+
+```python
+from graphene_django.views import GraphQLView
+from django.views.decorators.csrf import csrf_exempt
+from djql.schema import schema #change djql to your app name
+
+urlpatterns = [
+    path("graphql", csrf_exempt(GraphQLView.as_view(graphiql=True, schema=schema))),
+]
+```
+
+Graphql comes with an api browser similar to django's browsable api that you can use to test your queries and mutations. If you do not want to use it, you can set the `graphiql` parameter to False.
+The fourth import statement ```from djql.schema import schema``` is the schema that we will use to create our queries. Create a `schema.py` file in your project directory or your app directory.
+Django's csrf_exempt decorator is used to allow API clients to POST to the graphql endpoint we have created.
```

Comment:
```suggestion

The fourth `import` statement `from djql.schema import schema` is the schema that we will use to create our queries. Create a `schema.py` file in your project directory or your app directory.

Django's `csrf_exempt` decorator is used to allow API clients to POST to the GraphQL endpoint we have created.
```

"The _fourth_ import statement..."?? I only counted three?

Reply:
You're right, in my file there are 3. It's actually supposed to be 3.